Adds new file custom-preload that is loaded in a different order than custom.js before the main.js script is executed

lneves12 opened the pull request:
Adds new file custom-preload that is loaded in a different order than custom.js before the main.js script is executed.

Relying on the existing file custom.js has some limitation because it gets loaded after the main.js script is executed. That makes it hard to extend some behaviors like injecting security headers:

example:

```
      $.ajaxSetup({
        beforeSend: function (xhr) {
          xhr.setRequestHeader(
            "Authorization",
            `Bearer ${connectionDetails.password}`
          );
        },
      });
```

Here is a demo video demonstrating the new behavior:


https://user-images.githubusercontent.com/1140720/194337949-669faf59-e9f7-4888-9cc9-9be5317bf755.mp4

echarles commented:
@lneves12 Thx for opening this. The first thought while reviewing this is about the name you have chosen `custom` which is a bit too generic I believe. Apart from that, I am trying to understand the issue you are addressing and how you solved it could you expland a bit more. I get your point around loading order, but would need more background and details if possible. Thx again!

lneves12 commented:
Thanks for the quick reply @echarles 

I am not sure I understood the `custom` name comment. I don't think we have the control to change that, because that's the name of the module registered on `page.html`:

```
     // error-catching custom.js shim.
      define("custom", function (require, exports, module) {
          try {
              var custom = require('custom/custom');
              console.debug('loaded custom.js');
              return custom;
          } catch (e) {
              console.error("error loading custom.js", e);
              return {};
          }
      })
```

I am not an expert in require.js though. 

To fix this I kinda got inspired on this old PR: https://github.com/jupyter/notebook/pull/1732/files
That PR fixes the same issue on jupter V4, but it seems it never got backported to master. 


More about the use case: 

In order to use the iframe we need to extend some of the behavior. One of the examples (there are more thing) is to inject a couple of security headers. The problem is that the current `custom.js` is executed after `notebook.load_notebook(common_options.notebook_path);`.

Therefore doing things like:

```
      $.ajaxSetup({
        beforeSend: function (xhr) {
          xhr.setRequestHeader(
            "Authorization",
            `Bearer ${connectionDetails.password}`
          );
        },
      });
```

will not work, because not all the requests will have the injected token.

lneves12 commented:
Regarding the Linux JS Tests job failure, I am not sure yet what's going on, not sure if it's related.

All the tests are failing with `Message:   ReferenceError: Can't find variable: $` so it seems unrelated. 

I am trying to run the tests locally, but not luck yet

echarles commented:
> I am not sure I understood the custom name comment. I don't think we have the control to change that, because that's the name of the module registered on page.html:

You are right, thx for pointing to that content in the html page.

> In order to use the iframe we need to extend some of the behavior. One of the examples (there are more thing) is to inject a couple of security headers. The problem is that the current custom.js is executed after notebook.load_notebook(common_options.notebook_path);.

Could it be that other cases do need custom.js to be run after the notebook loading?

> Regarding the Linux JS Tests job failure, I am not sure yet what's going on, not sure if it's related. All the tests are failing with Message: ReferenceError: Can't find variable: $ so it seems unrelated.

The CI is not 100% green, but no test is failing with that error. I would say that it is related to these changes.

> python3 -m nbclassic.jstest notebook is failing with /usr/bin/python3: No module named nbclassic

You can follow the needed steps to run the tests in 

- https://github.com/jupyter/nbclassic/blob/3196dc6f8e1be0748662bfef11d397326b7ef0ea/.github/workflows/selenium.yml
- https://github.com/jupyter/nbclassic/blob/3196dc6f8e1be0748662bfef11d397326b7ef0ea/.github/workflows/js.yml

lneves12 commented:
> Could it be that other cases do need custom.js to be run after the notebook loading?
I would say no, in my opinion if there are cases like that they are probably badly written, because they can write code depending on jupyter triggered events (load, killed, etc).

btw this is how custom.js was being loaded in jupyter V4, it was just never backported to master. At least that's what i understood by reading the git history.   4.4.1 tag code: https://github.com/jupyter/notebook/blob/4.4.1/notebook/static/notebook/js/main.js#L27

The CI is not 100% green, but no test is failing with that error. I would say that it is related to these changes.

> You can follow the needed steps to run the tests in

thanks, I will try again to make them run locally

echarles commented:
> I would say no, in my opinion if there are cases like that they are probably badly written, because they can write code depending on jupyter triggered events (load, killed, etc).

We have just discussed that during the community call and we came to a conclusion that it would be preferable to not merge such change to avoid the risk to break any existing extensions.

> btw this is how custom.js was being loaded in jupyter V4, it was just never backported to master. At least that's what i understood by reading the git history. 4.4.1 tag code: https://github.com/jupyter/notebook/blob/4.4.1/notebook/static/notebook/js/main.js#L27

... we did not have that information while discussing during the community meeting. mmh would it be possible to hide that feature behind a flag on the command line. This will complexify quite a lot your PR, but it may be the safest way to move forward. WDYT?

lneves12 commented:
@echarles done! just updated the PR with the new file solution. I also updated the description with a new demo video showing the new file behaviors.

Can you rerun the CI tests again? I don't seem to have permissions to do that.

lneves12 commented:
Those seem to be failing as well in other PR: https://github.com/jupyter/nbclassic/actions/runs/3181760673/jobs/5186881765

There are 2 tests failing though in Selenium Tests only in Macos. I am still investigating why and how they can be related with my changes. 

Could they be flaky ? Or my change made them being more flaky? 

I will keep investigating, but any insight or help would be very welcoming! I am still very new to this project

Thanks!

echarles commented:
> Could they be flaky ? Or my change made them being more flaky?

Oh, maybe I have not looked correctly. Indeed, you just need to ensure that there are no regressions. We will work to make the CI 100% green, but for now we have some small failures that we know of. So if you could first double check that no regression is introduced, I can review.

echarles commented:
> @echarles could you try to rerun the tests? I rebased my branch.

Done

echarles commented:
Some tests were failing, so I have rerun the failing ones. They still show issues which I think were not there before, like on https://github.com/jupyter/nbclassic/actions/runs/3206644310/jobs/5260741971, there is 

```
Run python -m nbclassic.jstest notebook

Test group: notebook
Test file: /home/runner/work/nbclassic/nbclassic/nbclassic/tests/notebook/attachments.js
Timeout for http://localhost:8888/a@b/
Is the notebook server running?
FAIL ".CodeMirror-code" still did not exist in [1](https://github.com/jupyter/nbclassic/actions/runs/3206644310/jobs/5260741971#step:9:1)0000ms
```

lneves12 commented:
@echarles I think I see the same errors happening in last master commit.

jstests -> https://github.com/jupyter/nbclassic/actions/runs/3207336103/jobs/5242111041

selenium -> https://github.com/jupyter/nbclassic/actions/runs/3207336100/jobs/5242110926

=========================== short test summary info ============================
FAILED nbclassic/tests/selenium/test_save_as_notebook.py::test_save_notebook_as
FAILED nbclassic/tests/selenium/test_save_readonly_as.py::test_save_readonly_notebook_as
============ 2 failed, 29 passed, 821 warnings in 312.27s (0:05:12) ============

let me know if I overlooked something

echarles commented:
That's true. I'd like first to investigate what is wrong with the regression in main branch. How urgent is this for you?

lneves12 commented:
@echarles it's a blocker for us, so it has some priority. However I guess you will not release a new version until the tests are green. 

I would be surprise if this change breaks any test.

How can I help ? I can try to look at the js tests in master, but I haven't been able to run them locally yet, so it's a bit annoying to fix them. I will try again.

The selenium tests seem to be failing only in macos, not sure if I will be able to replicate them. (I have linux)

echarles commented:
I am investigating https://github.com/jupyter/nbclassic/pull/156 which is I think the cause of the test regression.

lneves12 commented:
thanks! @echarles  

let me know if there is anything I can do to help 

I don't mind helping fixing some of the tests and make the main branch green!

echarles commented:
https://github.com/jupyter/nbclassic/pull/159 on its way to fix the test in main branch. Once merged, you can rebase your PR.

PS: Sometimes it will be completely green, sometiimes you may have a few failures like `525 passed, 2 failed`.

echarles commented:
@lneves12 https://github.com/jupyter/nbclassic/pull/159 is merged in master. Can you please rebase this PR?

lneves12 commented:
@echarles nice!

My branch is rebased on top of latest master. You can retry the tests

echarles reviewed (approved):
LGTM Thx @lneves12

lneves12 commented:
@echarles thanks for all the help and quick replies! 

yeah, that would be very helpful. 

One question:  it seems only the notebook version > 6.5.X relies on nbclassic. Do you know if there is an official release of 6.5 happening soon?  or should we build our own docker images depending on the new release of nbclassic using the 6.5 branch? I am still new in the jupyter notebooks ecosystem, so any hint would be very helpful :) 

We have been using the official docker hub images: https://hub.docker.com/u/jupyter

echarles commented:
@lneves12 nbclassic 0.4.6 is just released with these changes.
